File: src/frontends/tensorflow_common/src/op/no_op.cpp
// Copyright (C) 2018-2025 Intel Corporation
// SPDX-License-Identifier: Apache-2.0
//

#include "common_op_table.hpp"

using namespace std;

namespace ov {
namespace frontend {
namespace tensorflow {
namespace op {

OutputVector translate_no_op(const NodeContext& node) {
    // the operation does nothing in terms of data generation
    default_op_checks(node, 0, {"NoOp", "SaveV2", "Assert", "LookupTableInsert", "LookupTableInsertV2"});
    return {};
}
}  // namespace op
}  // namespace tensorflow
}  // namespace frontend
}  // namespace ov

File: tools/mo/openvino/tools/mo/front/tf/noop.py
# Copyright (C) 2018-2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import logging as log

import networkx as nx

from openvino.tools.mo.front.common.replacement import FrontReplacementOp
from openvino.tools.mo.graph.graph import Graph
from openvino.tools.mo.utils.error import Error


class NoOpElimination(FrontReplacementOp):
    """
    NoOp does nothing and it has no data flow edges.
    It operates only with control flow edges.
    """
    op = "NoOp"
    enabled = True

    def replace_sub_graph(self, graph: Graph, match: dict):
        node = match['op']
        in_edges = node.in_edges()
        out_edges = node.out_edges()
        if len(in_edges) == 0 and len(out_edges) == 0:
            graph.remove_node(node.id)
            log.debug('NoOp op was removed {}'.format(node.id))
        else:
            raise Error('NoOp node {} contains data flow edges'.format(node.id))
